src/lib/hooks/useNotificationHandler.ts

```diff
@@ -0,0 +1,207 @@
+  const {currentConvoId} = useCurrentConvoId()
+  const {setShowLoggedOut} = useLoggedOutViewControls()
+  const closeAllActiveElements = useCloseAllActiveElements()
+
+  React.useEffect(() => {
+    const handleNotification = (payload?: NotificationRecord) => {
+      if (!payload) return
+
+      if (payload.reason === 'chat-message') {
+        if (payload.recipientDid !== currentAccount?.did && !storedPayload) {
+          storedPayload = payload
+          const account = accounts.find(a => a.did === payload.recipientDid)
+          if (account) {
+            onPressSwitchAccount(account, 'Notification')
+          } else {
+            closeAllActiveElements()
+            setShowLoggedOut(true)
+          }
+        } else {
+          setTimeout(() => {
+            // @ts-expect-error types are weird here
+            navigation.navigate('MessagesTab', {
+              screen: 'MessagesConversation',
+              params: {
+                conversation: payload.convoId,
+              },
+            })
+          }, 500)
+        }
+      } else {
+        switch (payload.reason) {
+          case 'like':
+          case 'repost':
+            resetToTab('NotificationsTab')
+            break
+          case 'follow':
+            const uri = new AtUri(payload.uri)
+            setTimeout(() => {
+              // @ts-expect-error types are weird here
+              navigation.navigate('HomeTab', {
+                screen: 'Profile',
+                params: {
+                  name: uri.host,
+                },
+              })
+            }, 500)
+            break
+          case 'mention':
+          case 'reply':
+            const urip = new AtUri(payload.uri)
+            setTimeout(() => {
+              // @ts-expect-error types are weird here
+              navigation.navigate('HomeTab', {
+                screen: 'PostThread',
+                params: {
+                  name: urip.host,
+                  rkey: urip.rkey,
+                },
+              })
+            }, 500)
+            break
+        }
+      }
+    }
+
+    Notifications.setNotificationHandler({
+      handleNotification: async e => {
+        if (e.request.trigger.type !== 'push') return DEFAULT_HANDLER_OPTIONS
+
+        logger.debug(
+          'Notifications: received',
+          {e},
+          logger.DebugContext.notifications,
+        )
+
+        const payload = e.request.trigger.payload as NotificationRecord
+        if (
+          payload.reason === 'chat-message' &&
+          payload.recipientDid === currentAccount?.did &&
+          currentConvoId !== payload.convoId
+        ) {
+          return {
+            shouldShowAlert: true,
+            shouldPlaySound: false,
+            shouldSetBadge: false,
+          }
+        }
+
+        // Any notification other than a chat message should invalidate the unread page
+        invalidateCachedUnreadPage()
+        return DEFAULT_HANDLER_OPTIONS
+      },
+    })
```

Comment:
This replaces the old `addNotificationReceivedListener`, because it lets us also control visibility of the push notification.